src/sources/prometheus/scrape.rs

```diff
@@ -529,6 +567,75 @@ mod test {
         }
     }
 
+    #[tokio::test]
+    async fn test_prometheus_request_query() {
+        let in_addr = next_addr();
+
+        let dummy_endpoint = warp::path!("metrics").and(warp::query::raw()).map(|query| {
+            format!(
+                r#"
+                    promhttp_metric_handler_requests_total{{query="{}"}} 100 1612411516789
+                "#,
+                query
+            )
+        });
+
+        tokio::spawn(warp::serve(dummy_endpoint).run(in_addr));
+
+        let query = HashMap::from([
+            ("key1".to_string(), vec!["val2".to_string()]),
+            (
+                "key2".to_string(),
+                vec!["val1".to_string(), "val2".to_string()],
+            ),
+        ]);
+
+        let config = PrometheusScrapeConfig {
+            endpoints: vec![format!("http://{}/metrics?key1=val1", in_addr)],
+            scrape_interval_secs: 1,
+            instance_tag: Some("instance".to_string()),
+            endpoint_tag: Some("endpoint".to_string()),
+            honor_labels: false,
+            query: Some(query.clone()),
+            auth: None,
+            tls: None,
+        };
+
+        let (tx, rx) = SourceSender::new_test();
+        let source = config
+            .build(SourceContext::new_test(tx, None))
+            .await
+            .unwrap();
+
+        tokio::spawn(source);
+        tokio::time::sleep(Duration::from_secs(1)).await;
+
+        let events = test_util::collect_ready(rx).await;
+        assert!(!events.is_empty());
+
+        let metrics: Vec<_> = events
+            .into_iter()
+            .map(|event| event.into_metric())
+            .collect();
+
+        for metric in metrics {
+            if let Some(q) = metric.tag_value("query") {
+                // note: request query has keys in arbitrary order
+                let mut parsed_query: HashMap<String, Vec<String>> = HashMap::new();
+                for (k, v) in url::form_urlencoded::parse(q.as_bytes()) {
+                    parsed_query
+                        .entry(k.to_string())
+                        .or_insert(Vec::new())
+                        .push(v.to_string());
+                }
+                for (k, l) in &query {
```

Comment:
I think this is failing to assert the query parameters encoded in `endpoint`. That is, we fail to test that there is `key1=val1` in here too.

My suggestion for refactoring this would be to:

* create the entire expected query as a hashmap like you did above on 585, just with the additional already encoded query parameter
* parse the received query into a hashmap
* just do `assert_eq` on the hashmaps

I see the comment about ordering, but I feel like ordering should be preserved using this approach. I could be wrong though.

Reply:
> I think this is failing to assert the query parameters encoded in `endpoint`. That is, we fail to test that there is `key1=val1` in here too.

Yes exactly. This is what I mentioned earlier that is not "checking the hashmap the other way around".



> My suggestion for refactoring this would be to:
> 
> * create the entire expected query as a hashmap like you did above on 585, just with the additional already encoded query parameter
> * parse the received query into a hashmap
> * just do `assert_eq` on the hashmaps

That was actually my very first idea to approach this test. However, HashMap does not implement `==` and thus I cannot simply assert that map1 == map2 unfortunately. This is why I resorted to comparing key-by-key.

> I see the comment about ordering, but I feel like ordering should be preserved using this approach. I could be wrong though.

The key ordering wouldn't matter if `==` was implemented and it compared the maps properly key-by-key. However this is (afaik) not possible without a helper function to do this comparison.

For now I will add the reverse comparison for completeness. Basically the code will look like this:
```rust
for (k, l) in &query {
    assert!(parsed_query.contains_key(k));
    // get the vectors of the key sorted properly like you suggested in another comment
    assert_eq!(parsed_key_vector, original_key_vector);
}
for (k, l) in &parsed_query {
    assert!(query.contains_key(k));
    // get the vectors of the key sorted properly like you suggested in another comment
    assert_eq!(parsed_key_vector, original_key_vector);
}
```

This will basically ensure that both maps are 1:1. It doesn't look very efficient nor elegant but I don't have any other better idea to do this :(

Reply:
Ah I see what I'm doing wrong. The `key1=val1` is not in the original query being sent, hence comparing these two maps 1:1 will never work 🤦. Indeed, as you say, I should construct an expected hashmap instead and compare to that.

Reply:
> * just do `assert_eq` on the hashmaps

I just went to re-test this again and it actually works indeed:
```rust
use std::collections::HashMap;

fn main() {
    let expected = HashMap::from([
        (
            "key1".to_string(),
            vec!["val1".to_string(), "val2".to_string()],
        ),
        (
            "key2".to_string(),
            vec!["val1".to_string(), "val2".to_string()],
        ),
    ]);

    let got = HashMap::from([
        (
            "bad-key".to_string(),
            vec!["val1".to_string(), "val2".to_string()],
        ),
        (
            "key2".to_string(),
            vec!["val1".to_string(), "val2".to_string()],
        ),
    ]);

    assert_eq!(expected, got);
}

// Compiling playground v0.0.1 (/playground)
//    Finished dev [unoptimized + debuginfo] target(s) in 5.05s
//     Running `target/debug/playground`
// thread 'main' panicked at 'assertion failed: `(left == right)`
//   left: `{"key1": ["val1", "val2"], "key2": ["val1", "val2"]}`,
//  right: `{"key2": ["val1", "val2"], "bad-key": ["val1", "val2"]}`', src/main.rs:26:5
// note: [run with `RUST_BACKTRACE=1` environment variable to display a backtrace](https://play.rust-lang.org/#)
```

I swear I saw the compiler before complaining that `==` was not implemented for hashmaps :/
Perhaps when I implemented it initially, I did something wrong before without noticing.

Anyway, I will then refactor the test to be much simpler using this approach 👍.

Reply:
Awesome! Yeah, `HashMap` will implement `Eq` so long as the values and keys do: https://doc.rust-lang.org/std/collections/struct.HashMap.html#impl-Eq

> impl<K, V, S> [Eq](https://doc.rust-lang.org/std/cmp/trait.Eq.html) for [HashMap](https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html)<K, V, S>
> where
>     K: [Eq](https://doc.rust-lang.org/std/cmp/trait.Eq.html) + [Hash](https://doc.rust-lang.org/std/hash/trait.Hash.html),
>     V: [Eq](https://doc.rust-lang.org/std/cmp/trait.Eq.html),
>     S: [BuildHasher](https://doc.rust-lang.org/std/hash/trait.BuildHasher.html),

Reply:
Refactored the test now as discussed, got way simpler indeed!